optax/contrib/_dog.py

```diff
@@ -79,37 +83,87 @@ def init_fn(params: base.Params) -> DoGState:
     if init_step_type == "distance":
       r_epsilon = init_step_value
     elif init_step_type == "heuristic":
-      r_epsilon = init_step_value * (1 + optax.tree.norm(params))
+        if layer_wise:
+            r_epsilon = jax.tree.map(
+                lambda p: init_step_value * (1 + numerics.safe_norm(p, 0.0)),
+                params,
+            )
+        else:
+            r_epsilon = init_step_value * (1 + optax.tree.norm(params))
     elif init_step_type == "learning_rate":
       r_epsilon = 0.0
     else:
       raise ValueError(
           f"Invalid init_step specification for scale_by_dog: {init_step_type=}"
       )
 
+    if layer_wise:
+        if init_step_type == "heuristic":
+             # r_epsilon is already a tree of scalars
+             max_dist = jax.tree.map(
+                 lambda r: jnp.asarray(r, dtype=params_dtype), r_epsilon
+             )
+        else:
+             # r_epsilon is a scalar, create tree of scalars
+             max_dist = jax.tree.map(
+                 lambda p: jnp.full([], r_epsilon, dtype=params_dtype), params
+             )
+        sum_sq_norm_grads = jax.tree.map(
+            lambda p: jnp.zeros([], dtype=params_dtype), params
+        )
+    else:
+        max_dist = jnp.asarray(r_epsilon, dtype=params_dtype)
+        sum_sq_norm_grads = jnp.asarray(0.0, dtype=params_dtype)
+
     return DoGState(
         is_init_step=jnp.asarray(True),
         init_params=params,
-        max_dist=jnp.asarray(r_epsilon, dtype=params_dtype),
-        sum_sq_norm_grads=jnp.asarray(0.0, dtype=params_dtype),
+        max_dist=max_dist,
+        sum_sq_norm_grads=sum_sq_norm_grads,
     )
 
   def update_fn(
       updates: base.Updates, state: DoGState, params: base.Params
   ) -> tuple[base.Updates, DoGState]:
-    dist = optax.tree.norm(optax.tree.sub(state.init_params, params))
-    max_dist = jnp.maximum(state.max_dist, dist)
-    sum_sq_norm_grads = state.sum_sq_norm_grads + optax.tree.norm(
-        updates, squared=True
-    )
-    learning_rate = max_dist / jnp.sqrt(sum_sq_norm_grads + eps)
 
-    if init_step_type == "learning_rate":
-      learning_rate = jnp.where(
-          state.is_init_step, init_step_value, learning_rate
-      )
+    if layer_wise:
```

Comment:
This function body is almost completely different depending on `layer_wise`. Could you factor out two helper functions, and call one or the other depending on the value of `layer_wise`? 

(Also, if I understand right, the motivation for this PR was to reduce repeated code, but it seems like ultimately there wasn't that much repeated code?)